File: benchmarks/go/vector_add.go
//go:build ignore

// Vector addition benchmark - Go
// Run: go run vector_add.go
package main

import (
	"fmt"
	"time"
)

const N = 1024
const ITERATIONS = 1000000

//go:noinline
func vectorAddSum(a, b, c []float32) float32 {
	for i := range a {
		c[i] = a[i] + b[i]
	}
	var sum float32
	for _, v := range c {
		sum += v
	}
	return sum
}

func main() {
	a := make([]float32, N)
	b := make([]float32, N)
	c := make([]float32, N)

	// Initialize
	for i := 0; i < N; i++ {
		a[i] = float32(i%100) * 0.01
		b[i] = float32(i%50) * 0.02
	}

	start := time.Now()

	var result float32
	for iter := 0; iter < ITERATIONS; iter++ {
		result = vectorAddSum(a, b, c)
	}

	elapsed := time.Since(start)

	fmt.Printf("Go vector_add: %.3f ms (%d iterations, N=%d)\n",
		float64(elapsed.Milliseconds()), ITERATIONS, N)
	fmt.Printf("Checksum: %.2f\n", result)
	fmt.Printf("Throughput: %.2f M elements/sec\n",
		float64(ITERATIONS)*N/elapsed.Seconds()/1e6)
}

File: benchmarks/c/vector_add.c
// Vector addition benchmark - C with AVX2
// Compile: gcc -O3 -mavx2 -o vector_add vector_add.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <immintrin.h>

#define N 1024
#define ITERATIONS 1000000

__attribute__((noinline))
float vector_add_sum(float* a, float* b, float* c, int n) {
    // AVX2 vectorized addition
    int i;
    for (i = 0; i + 8 <= n; i += 8) {
        __m256 va = _mm256_loadu_ps(&a[i]);
        __m256 vb = _mm256_loadu_ps(&b[i]);
        __m256 vc = _mm256_add_ps(va, vb);
        _mm256_storeu_ps(&c[i], vc);
    }
    // Scalar cleanup
    for (; i < n; i++) {
        c[i] = a[i] + b[i];
    }

    // Sum reduction
    float sum = 0.0f;
    for (i = 0; i < n; i++) {
        sum += c[i];
    }
    return sum;
}

int main() {
    float *a = aligned_alloc(32, N * sizeof(float));
    float *b = aligned_alloc(32, N * sizeof(float));
    float *c = aligned_alloc(32, N * sizeof(float));

    // Initialize
    for (int i = 0; i < N; i++) {
        a[i] = (float)(i % 100) * 0.01f;
        b[i] = (float)(i % 50) * 0.02f;
    }

    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);

    volatile float result = 0;
    for (int iter = 0; iter < ITERATIONS; iter++) {
        result = vector_add_sum(a, b, c, N);
    }

    clock_gettime(CLOCK_MONOTONIC, &end);

    double elapsed = (end.tv_sec - start.tv_sec) +
                     (end.tv_nsec - start.tv_nsec) / 1e9;

    printf("C AVX2 vector_add: %.3f ms (%d iterations, N=%d)\n",
           elapsed * 1000, ITERATIONS, N);
    printf("Checksum: %.2f\n", result);
    printf("Throughput: %.2f M elements/sec\n",
           (double)ITERATIONS * N / elapsed / 1e6);

    free(a);
    free(b);
    free(c);
    return 0;
}
